### DISCUSSION TITLE
Mapping actions to stylus and tablet buttons?

### DISCUSSION THREAD

**Discussion by kzembower** [xournalpp/xournalpp#3296] [Q&A] [answered]
Hello, all,

According to the README.md file for version 1.1.0, features include "Allows mapping different tools/colors etc. to stylus/mouse buttons."

Can someone point me to where this is documented? I read the current User Manual, but didn't see it. Has this feature not been documented yet? If so, if someone can help me get started, I'll edit the documentation to include it.

Thanks.

-kevin

**Comment by personalizedrefrigerator** [comment, marked as answer]
See `Edit > Preferences > Stylus` and `Edit > Preferences > Mouse`:

![image](https://user-images.githubusercontent.com/46334387/128462545-c77ca4f1-525d-4aa3-be85-334a5a7bc433.png)

I'm not sure if it's been documented yet...

**Edit**: Fixed typo.

**Comment by kzembower** [comment]
Hello, Henry, thanks for responding.

My copy of version 1.1.0 (git commit 9d1277de) doesn't have Edit -> 
Settings. It has Edit -> Preferences -> Mouse and -> Stylus; is this 
what you're thinking of?

I guess this technically answers the question of where to make these 
changes, but I'm left dissatisfied and confused. I could have sworn that 
previous versions of xournal++ had these features already. Also, I was 
hoping for a way to use the 12 buttons on my Huion 1060PLUS tablet in 
xournal++, but it seems that this still can't be accomplished.

Thanks, again, for your help and suggestion.

-Kevin

**Reply by rolandlo** [reply]
@kzembower You can use your 12 buttons using the [plugin functionality](https://xournalpp.github.io/guide/plugins/plugins/). Just map each button to some (yet unused) keyboard accelerator and act upon them in the plugin. Ask further questions if needed.
And yes, `Settings` instead of `Preferences` was certainly a typo and the preference settings pointed out by @personalizedrefrigerator were already there in previous versions. There is no claim in the list of features in the README, that these features have all only been introduced for version 1.1.0. Some are new, some have been refined, some have existed before. If you want to learn what is new in version 1.1.0 you should read the [changelog](https://github.com/xournalpp/xournalpp/blob/master/CHANGELOG.md).

**Reply by personalizedrefrigerator** [reply]
> It has Edit -> Preferences -> Mouse and -> Stylus; is this what you're thinking of?

Yes. Thank you for the correction.

> I could have sworn that previous versions of xournal++ had these features already

I think they did... <strike>I'm confused as well.</strike> **Edit:** Thank you for clarifying @rolando!

> Also, I was
hoping for a way to use the 12 buttons on my Huion 1060PLUS tablet in xournal++

The solution mentioned here might help:
https://krita-artists.org/t/hotkeys-on-huion-tablet/3222/6

The above, however, is for all applications. If you want actions specific to Xournal++, I suggest that you:
 * Map a button to a keystroke
   * See answers 6 and 7 to the question linked above.
 * Write a Lua plug-in that does something when it receives that keystroke.
   * See https://xournalpp.github.io/guide/plugins/plugins/

**Comment by gapato** [comment]
This thread is very old, but in case someone is interested, I created a [small plugin](https://github.com/gapato/xournalpp_keypad2colors) to map the keypad keys to colors, which can serve as an example.

By the way, I don't know lua, so the (small) code is full of copy-paste, any advice on making it better would be welcome!

**Reply by rolandlo** [reply]
I would probably add all data to a table:
```lua
local COLORS = { 
  { hex = 0x0000000, desc = "Color 0", acc="KP_0"},
  { hex = 0x0008000, desc = "Color 1", acc="KP_1"},
    ...
}
```
and then you can use just one callback in different modes:
```lua
function initUi()
  for i, col in ipairs(COLORS) do
    app.registerUi({ ["menu"] = col.desc, ["callback"] = "switchColor", ["accelerator"] = col.acc, ["mode"] = i })
  end
end

function switchColor(mode)
  app.changeToolColor({ ["color"] = COLORS[mode].hex }) 
end
```

Of course when the color descriptions are just "Color 0", "Color 1" etc. and the accelerators are just "KP_0", "KP_1" etc. then you could also compute `col.desc` and `col.acc` from the index i (resp. i-1). Then there would be no need to use a more complicated table than the one you have in the code.

**Reply by gapato** [reply]
Thank you for your answer!

I tried a while, blindly, to pass arguments to the callback, unsuccessfully. With this I can make it much cleaner.
